### PR TITLE
fix(core, web): fix compatibility with TrustedTypes

### LINKED ISSUE #12301: 🐛 [firebase_core_web] TrustedType violations in firebase_core_web.dart - TrustedScript needed
## Bug report

Several Google Flutter web apps using Firebase plugins that rely on firebase_core_web are receiving browser error reports through the [TrustedTypes](https://web.dev/trusted-types/) browser APIs. Specifically, the script injection code in [_injectSrcScript](https://github.com/firebase/flutterfire/blob/master/packages/firebase_core/firebase_core_web/lib/src/firebase_core_web.dart) is not adhering to the standards and is causing errors to be logged.

We really want to get this fixed so that we can enable new security features which will protect our apps from certain types of vulnerability.

This was previously reported in https://github.com/firebase/flutterfire/issues/10311 and a fix attempted in https://github.com/firebase/flutterfire/pull/10312, however the fix didn't work (see below). It would be amazing if you could try to reproduce the issue with steps below and verify that using a TrustedScript fixes it.

Currently, we have code like:
```
    web.TrustedScriptURL? trustedUrl;
    final trustedTypePolicyName = _defaultTrustedPolicyName + windowVar;
    if (web.window.nullableTrustedTypes != null) {
      web.console.debug(
        'TrustedTypes available. Creating policy: $trustedTypePolicyName'.toJS,
      );
      try {
        final web.TrustedTypePolicy policy =
            web.window.trustedTypes.createPolicy(
          trustedTypePolicyName,
          web.TrustedTypePolicyOptions(
            createScriptURL: ((JSString url) => src).toJS,
          ),
        );
        trustedUrl = policy.createScriptURLNoArgs(src);
      } catch (e) {
        throw TrustedTypesException(e.toString());
      }
    }

    final web.HTMLScriptElement script =
        web.document.createElement('script') as web.HTMLScriptElement;
    script.type = 'text/javascript';
    script.crossOrigin = 'anonymous';
    final stringUrl = trustedUrl != null
        // Necessary for the JS interop to work correctly on Flutter Beta 3.19.
        // ignore: unnecessary_cast
        ? (trustedUrl as JSObject).callMethod('toString'.toJS)
        : src;
    script.text = '''
      window.ff_trigger_$windowVar = async (callback) => {
        console.debug("Initializing Firebase $windowVar");
        callback(await import("$stringUrl"));
      };
    ''';
```

The issue here is that although we do create a trusted types policy to create a TrustedScriptURL (added in the previous fix), we're still "unsafely" setting the text of the Script element (e.g. dynamically creating an inline script) rather than its URL. So even though we embed the DomTrustedScriptUrl in the inline script, overall the entire inner javascript is untrusted. It seems like we should actually use a TrustedScript policy here to create a trusted script to embed - see https://developer.mozilla.org/en-US/docs/Web/API/TrustedScript.

Also see related issue https://github.com/firebase/firebase-js-sdk/issues/7048 which was fixed.

### Steps to reproduce the behavior:

1. Create an example Flutter web app using Firebase
2. Enable trusted types enforcement, e.g. by adding:

    ```  
    <meta
      http-equiv="Content-Security-Policy"
      content="require-trusted-types-for 'script'"
     />
    ```
3. Load page, and look for errors in the JS console

### Expected behavior

Setting the text on the script tag should not produce a trusted types error.

### PR DESCRIPTION
**Pull request by Lyokone**
## Description

*Replace this paragraph with a description of what this PR is doing. If you're modifying existing behavior, describe the existing behavior, how this PR is changing it, and what motivated the change.*

## Related Issues

closes #12301 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I signed the [CLA].
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/firebase_core/firebase_core/example/web/index.html [hunk 1]
```diff
@@ -19,10 +19,18 @@
     <meta charset="UTF-8" />
     <meta content="IE=Edge" http-equiv="X-UA-Compatible" />
 
-    <meta
+    <!-- 
+      If you wanna test the require-trusted-types-for CSP, uncomment the line below
+      and run the app in a browser that supports Trusted Types (e.g. Chrome 83+).
+      Note that you need to run the app in release mode till this issue is fixed:
+      https://github.com/requirejs/requirejs/issues/1832
```
Comment (russellwheatley) [inline packages/firebase_core/firebase_core/example/web/index.html:26]:
```suggestion
    <!-- 
      If you want to test the require-trusted-types-for CSP, uncomment the line below
      and run the app in a browser that supports Trusted Types (e.g. Chrome 83+).
      Note that you need to run the app in release mode until this issue is fixed:
      https://github.com/requirejs/requirejs/issues/1832
```

**Review by russellwheatley — APPROVED** [review]
Made a suggestion to docs, other than that, LGTM

File: packages/firebase_core/firebase_core/example/web/index.html [hunk 2]
```diff
@@ -19,10 +19,18 @@
     <meta charset="UTF-8" />
     <meta content="IE=Edge" http-equiv="X-UA-Compatible" />
 
-    <meta
+    <!-- 
+      If you wanna test the require-trusted-types-for CSP, uncomment the line below
+      and run the app in a browser that supports Trusted Types (e.g. Chrome 83+).
+      Note that you need to run the app in release mode till this issue is fixed:
```
Comment (Salakar) [inline packages/firebase_core/firebase_core/example/web/index.html:25]:
```suggestion
      Note that you need to run the app in release mode until this issue is fixed:
```

Comment (Lyokone) [reply packages/firebase_core/firebase_core/example/web/index.html]: Applied

File: packages/firebase_core/firebase_core/example/web/index.html [hunk 3]
```diff
@@ -19,10 +19,18 @@
     <meta charset="UTF-8" />
     <meta content="IE=Edge" http-equiv="X-UA-Compatible" />
 
-    <meta
+    <!-- 
+      If you wanna test the require-trusted-types-for CSP, uncomment the line below
```
[the remark this answers is not in the document]
Comment (Lyokone) [reply packages/firebase_core/firebase_core/example/web/index.html]: Fixed

**Review by exaby73 — APPROVED** [review]

**Comment by nilsreichardt** [comment]
A new version of `web` has been released: `v0.5.1`. It contains support for `TrustedTypes`. Should replace this PR with the version of `web`?

**Pull request merged**